pkg/scheduler/plugins/gang/plugin.go

```diff
@@ -0,0 +1,534 @@
+// i. handle the timeout gang
+// ii. do nothing when bound failed
+func (p *GangPlugin) Unreserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) {
+	gangName := getGangNameByPod(pod)
+	if gangName == "" {
+		return
+	}
+	gangId := getNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := p.gangCache.getGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName: %v, podName: %v", gangId,
+			getNamespaceSplicingName(pod.Namespace, pod.Name))
+		return
+	}
+
+	// gang time out
+	if gang.isGangTimeout() {
+		klog.Infof("gang is time out,start to release the assumed resource and add annotations to the "+
+			"gang's children, gangName: %v, podName: %v", gangId, getNamespaceSplicingName(pod.Namespace, pod.Name))
+		// gang is timeout,we clear the TimeoutStartTime
+		gang.clearTimeoutStartTime()
+		pods, err := p.podLister.List(labels.NewSelector())
+		if err != nil {
+			klog.Errorf("unReserve list pod err: %v", err.Error())
+			return
+		}
+		for _, podOriginal := range pods {
+			if getNamespaceSplicingName(pod.Namespace,
+				getGangNameByPod(pod)) == gangId {
+				pod = podOriginal.DeepCopy()
+				if pod.Annotations == nil {
+					pod.Annotations = map[string]string{}
+				}
+				pod.Annotations[extension.AnnotationGangTimeout] = "true"
+				patchBytes, err := generatePodPatch(podOriginal, pod)
+				if err != nil {
+					klog.Errorf("unReserve generatePodPatch got err: %v", err.Error())
+					continue
+				}
+				if string(patchBytes) == "{}" {
+					continue
+				}
+				err = retry.OnError(
+					retry.DefaultRetry,
+					errors.IsTooManyRequests,
+					func() error {
+						_, err = p.frameworkHandle.ClientSet().CoreV1().Pods(pod.Namespace).Patch(ctx, pod.Name,
+							types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+						return err
+					})
+				if err != nil {
+					klog.Errorf("Failed to patch gang timeout annotation after retry to pod: %v, gang: %v, err: %v",
+						getNamespaceSplicingName(pod.Namespace, pod.Name), gangId, err.Error())
+				} else {
+					klog.Infof("unReserve patch gang timeout annotation to pod: %v, gang: %v success",
+						getNamespaceSplicingName(pod.Namespace, pod.Name), gangId)
+				}
+			}
+		}
+
+		// release resource of all assumed children of the gang
+		p.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+			if getNamespaceSplicingName(waitingPod.GetPod().Namespace,
+				getGangNameByPod(waitingPod.GetPod())) == gangId {
+				klog.Errorf("unReserve rejects the pod name: %v from Gang: %s due to timeout",
+					getNamespaceSplicingName(pod.Namespace, pod.Name), gangId)
+				waitingPod.Reject(p.Name(), "optimistic rejection in unReserve due to timeout")
+			}
+		})
+	}
+}
+
+// PostBind just update the gang's BoundChildren
+func (p *GangPlugin) PostBind(ctx context.Context, _ *framework.CycleState, pod *corev1.Pod, nodeName string) {
+	gangName := getGangNameByPod(pod)
+	if gangName == "" {
```

Comment:
if the gang is PodGroup, should update the PodGroup status..

Reply:
yes